### PR TITLE
Experimental MUSL target support for the compiler

### PR DESCRIPTION
**Pull request by alexcrichton**
These commits build on [some great work on reddit](http://www.reddit.com/r/rust/comments/33boew/weekend_experiment_link_rust_programs_against/) for adding MUSL support to the compiler. This goal of this PR is to enable a `--target x86_64-unknown-linux-musl` argument to the compiler to work A-OK. The outcome here is that there are 0 compile-time dependencies for a MUSL-targeting build _except for a linker_. Currently this also assumes that MUSL is being used for statically linked binaries so there is no support for dynamically linked binaries with MUSL.

MUSL support largely just entailed munging around with the linker and where libs are located, and the major highlights are:
- The entirety of `libc.a` is included in `liblibc.rlib` (statically included as an archive).
- The entirety of `libunwind.a` is included in `libstd.rlib` (like with liblibc).
- The target specification for MUSL passes a number of ... flavorful options! Each option is documented in the relevant commit.
- The entire test suite currently passes with MUSL as a target, except for:
  - Dynamic linking tests are all ignored as it's not supported with MUSL
  - Stack overflow detection is not working MUSL yet (I'm not sure why)
- There is a language change included in this PR to add a `target_env` `#[cfg]` directive. This is used to conditionally build code for only MUSL (or for linux distros not MUSL). I highly suspect that this will also be used by Windows to target MSVC instead of a MinGW-based toolchain.

To build a compiler targeting MUSL you need to follow these steps:
1. Clone the current MUSL repo from `git://git.musl-libc.org/musl`. Build this as usual and install it.
2. Clone and build LLVM's [libcxxabi](http://libcxxabi.llvm.org/) library. Only the `libunwind.a` artifact is needed. I have tried using upstream libunwind's source repo but I have not gotten unwinding to work with it unfortunately. Move `libunwind.a` adjacent to MUSL's `libc.a`
3. Configure a Rust checkout with `--target=x86_64-unknown-linux-musl --musl-root=$MUSL_ROOT` where `MUSL_ROOT` is where you installed MUSL in step 1.

I hope to improve building a copy of libunwind as it's still a little sketchy and difficult to do today, but other than that everything should "just work"! This PR is not intended to include 100% comprehensive support for MUSL, as future modifications will probably be necessary.

### CODE REVIEW

**Comment by brson** [comment]
Why 'x86_64-unknown-linux-musl'? It seems reasonable to me, but Google doesn't have a lot of hits for it.

**Comment by nikomatsakis** [comment]
Neat!

**Comment by alexcrichton** [comment]
@brson I found it followed [clang's target triple format](http://clang.llvm.org/docs/CrossCompilation.html#general-cross-compilation-options-in-clang), but I personally got the idea from the [reddit post](https://gist.github.com/cl91/bb927df2525738502131#the-plan).

**Comment by brson** [comment]
r=me

**Comment by alexcrichton** [comment]
@brson I've pushed some more commits clarifying some documentation and also feature gating `target_env` by just not defining it unless the gate is turned on. I've avoided rebasing for now but I'll rebase all the commits up after review.

**Comment by aidanhs** [comment]
@alexcrichton got any details of the issues you had with libunwind upstream?

**Comment by alexcrichton** [comment]
@aidanhs unfortunately I could never get any more detail other than "unwinding didn't work". I was able to build libunwind upstream with `musl-gcc` (aka build it against MUSL's libc.a), and it was also correctly linked into the standard library and Rust programs. When running, however, it either couldn't find the unwind tables or couldn't find the handler for the unwind, as all attempts to unwind failed with "handler not found".

In your test program, did you trigger an unwind, or was it just running as usual? Also, did you configure libunwind with any special flags or anything?

File: src/libstd/sys/unix/thread.rs [hunk 1]
```diff
@@ -342,7 +342,10 @@ fn min_stack_size(attr: *const libc::pthread_attr_t) -> usize {
     static mut __pthread_get_minstack: Option<F> = None;
 
     INIT.call_once(|| {
-        let lib = DynamicLibrary::open(None).unwrap();
+        let lib = match DynamicLibrary::open(None) {
+            Ok(l) => l,
+            Err(..) => return,
+        };
```
Comment (nagisa) [inline src/libstd/sys/unix/thread.rs]: This is why stack overflow detection is not working.

**Comment by aidanhs** [comment]
@alexcrichton oops, I didn't get unwinding working (though normal execution worked). I did get a different error to you though:
```
$ ./target/debug/binary
thread '<main>' panicked at 'mypanic', src/main.rs:28
fatal runtime error: Could not unwind stack, error = 5
Illegal instruction (core dumped)
```

Sorry for the noise!

File: src/libstd/sys/unix/thread.rs (hunk 1 above)
Comment (alexcrichton) [inline src/libstd/sys/unix/thread.rs]: Hm, could you elaborate a little more? This should in theory just affect the return value of this function, which I'm not sure how connected it is to stack sizes.

**Comment by alexcrichton** [comment]
> I did get a different error to you though:

Ah that's actually the same error that I was seeing, oh well!

File: src/libstd/sys/unix/thread.rs (hunk 1 above)
Comment (nagisa) [inline src/libstd/sys/unix/thread.rs]: Before it failed loud and clear when this didn’t work, but now `min_stack_size` returns `PTHREAD_STACK_MIN` instead which might or might not be a correct value to use in the stack size calculations.

/me shrugs.

**Comment by jsarenik** [comment]
Is there any update with this, please? I am getting `Illegal instruction` on Alpine Linux x86_64.

```
$ uname -mrv
5.10.27-0-lts #1-Alpine SMP Wed, 31 Mar 2021 06:03:59 +0000 x86_64
$ rustc --version
rustc 1.51.0
```

The freshly and successfully compiled binary is of [`electrs`](https://github.com/romanz/electrs/blob/master/doc/usage.md#static-linking).

**Pull request merged**